ndk/src/midi.rs

```diff
@@ -0,0 +1,271 @@
+//! Bindings for [`AMidiDevice`], [`AMidiInputPort`], and [`AMidiOutputPort`]
+//!
+//! See [the NDK guide](https://developer.android.com/ndk/guides/audio/midi) for
+//! design and usage instructions, and [the NDK reference](https://developer.android.com/ndk/reference/group/midi)
+//! for an API overview.
+//!
+//! [`AMidiDevice`]: https://developer.android.com/ndk/reference/group/midi#amididevice
+//! [`AMidiInputPort`]: https://developer.android.com/ndk/reference/group/midi#amidiinputport
+//! [`AMidiOutputPort`]: https://developer.android.com/ndk/reference/group/midi#amidioutputport
+#![cfg(feature = "media")]
+
+pub use super::media::Result;
+use super::media::{construct, NdkMediaError};
+
+use num_enum::{IntoPrimitive, TryFromPrimitive};
+use std::fmt;
+use std::marker::PhantomData;
+use std::os::raw::{c_int, c_uint};
+use std::ptr::NonNull;
+
+/// Result of [`MidiOutputPort::receive`].
+#[derive(Copy, Clone, Debug)]
+#[repr(u32)]
+pub enum MidiOpcode {
+    /// No MIDI messages are available.
+    NoMessage,
+    /// Received a MIDI message with the given length.
+    Data(usize),
+    /// Instructed to discard all pending MIDI data.
+    Flush,
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, TryFromPrimitive, IntoPrimitive)]
+#[repr(u32)]
+pub enum MidiDeviceType {
+    Bluetooth = ffi::AMIDI_DEVICE_TYPE_BLUETOOTH,
+    USB = ffi::AMIDI_DEVICE_TYPE_USB,
+    Virtual = ffi::AMIDI_DEVICE_TYPE_VIRTUAL,
+}
+
+#[derive(Debug)]
+pub struct MidiDevice {
+    inner: NonNull<ffi::AMidiDevice>,
+}
+
+impl MidiDevice {
+    fn from_ptr(inner: NonNull<ffi::AMidiDevice>) -> Self {
+        Self { inner }
+    }
+
+    fn as_ptr(&self) -> *mut ffi::AMidiDevice {
+        self.inner.as_ptr()
+    }
```

Comment:
Is this intentionally private? Looks like a copy-paste from `image_reader.rs`, all the other wrappers use a `pub unsafe` with some documentation explaining if ownership is taken or not.

Reply:
I referred `audio.rs`, but seems like `audio.rs` also does not comply with the rules you mentioned. I'll make a fix for this.

Reply:
I changed `from_ptr` into a `pub unsafe` function, do I have to do the same to functions in `audio.rs`?

Reply:
Fwiw we don't use the ``Creates an `MidiDevice` from a pointer.`` language anywhere else. Since this struct takes/inherits ownership of the pointer (it doesn't _create_ anything, these structs are merely thin wrappers around a pointer), also by closing/releasing these objects on `Drop`, can you use the same ``Assumes ownership of `ptr` `` terminology I used for `NativeWindow`/`SurfaceTexture`?

Also, it's probably fine to make `fn ptr()` public.

Reply:
![image](https://user-images.githubusercontent.com/17005454/194041047-9cbf0ff6-e756-4b0b-ae14-899c792be899.png)
Fixed!

Reply:
Probably no need to have _both_ `fn ptr()` _and_ `as_ptr()`?

It looks like we have more of the former, probably stick with that.

Reply:
Fixed!